zookeeper-server/src/test/java/org/apache/zookeeper/log/Log4jConfigWatcherTest.java

```diff
@@ -0,0 +1,126 @@
+
+    @Before
+    public void setUp() throws Exception {
+        File tempDir = ClientBase.createEmptyTestDir();
+        testLogFilePath = tempDir.getPath() + "/zookeeper.log";
+        System.setProperty(LOG_4_J_WATCHER_LOG_DIR, tempDir.getPath());
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        System.clearProperty(Log4jConfigWatcher.LOG4J_CONFIGURATION_WATCH);
+        System.clearProperty(Log4jConfigWatcher.LOG4J_CONFIGURATION);
+        System.clearProperty(Log4jConfigWatcher.LOG4J_CONFIGURATION_WATCH_INTERVAL);
+        System.clearProperty(LOG_4_J_WATCHER_LOG_DIR);
+        File dir = new File(testLogFilePath);
+        dir.delete();
+    }
+
+    @Test
+    public void testLogLevelChangeWithoutRestartingJVM() throws Exception {
+        System.setProperty(Log4jConfigWatcher.LOG4J_CONFIGURATION_WATCH, "true");
+        System
+            .setProperty(Log4jConfigWatcher.LOG4J_CONFIGURATION, "log4j-watcher-log4j.properties");
+        System.setProperty(Log4jConfigWatcher.LOG4J_CONFIGURATION_WATCH_INTERVAL, "2000");
+        Log4jConfigWatcher.watchLog4jConfiguration();
+
+        // Changed the log level to WARN
+        alterLogConfigFile("/" + System.getProperty(Log4jConfigWatcher.LOG4J_CONFIGURATION),
+            zookeeperLogLevel, zookeeperLogLevel + "WARN, ROLLINGFILE");
+        // Waiting for some time so configuration event is triggered and changes reloaded into memory
+        Thread.sleep(5000);
+        Logger LOG = LoggerFactory.getLogger(Log4jConfigWatcherTest.class);
+        LOG.warn("Warn level log is present in initial Configuration");
+        LOG.info("Info level log is present in initial Configuration");
+        Assert.assertTrue(checkTargetInLogFile(testLogFilePath,
+            "Warn level log is present in initial Configuration"));
+        Assert.assertFalse(checkTargetInLogFile(testLogFilePath,
+            "Info level log is present in initial Configuration"));
+
+        // Changed the log level to INFO
+        alterLogConfigFile("/" + System.getProperty(Log4jConfigWatcher.LOG4J_CONFIGURATION),
+            zookeeperLogLevel, zookeeperLogLevel + "INFO, ROLLINGFILE");
+        // Waiting for some time so configuration event is triggered and changes reloaded into memory
+        Thread.sleep(5000);
+        LOG.info("Info level log is present after altering the Configuration");
+        Assert.assertTrue(checkTargetInLogFile(testLogFilePath,
+            "Info level log is present after altering the Configuration"));
+    }
+
+    private void alterLogConfigFile(String log4jConfigurationFile, String key, String value)
+        throws IOException {
+        InputStream inputStream = null;
+        BufferedWriter bw = null;
+        try {
+            // log4j configuration file is modified from classpath, not from the resource folder
+            inputStream = getClass().getResourceAsStream(log4jConfigurationFile);
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [UI: Usage of GetResource may be unsafe if class is extended (UI_INHERITANCE_UNSAFE_GETRESOURCE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#ui-usage-of-getresource-may-be-unsafe-if-class-is-extended-ui-inheritance-unsafe-getresource).
Calling this.getClass().getResource(...) could give results other than expected if this class is extended by a class in another package.